### PR TITLE
bpf: Introduce per-cluster NAT maps

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
This is a part of the Cluster Mesh with overlapping PodCIDR support. When the remote cluster has an overlapping PodCIDR, we must identify the remote endpoint using IP + ClusterID. NAT is not an exception, so it's ideal for extending the NAT map's key with a new ClusterID field. However, the problem is we cannot extend the NAT map without breaking users' connections now. Thus, we created a new map-in-map containing the NAT maps for each cluster.

This PR only contains basic map definitions and tests. No one uses it. We also don't have GC for per-cluster maps. Once #22857 and this PR are merged, I'll submit the PR for GCs.

```release-note
clustermesh: Introduce per-cluster NAT maps
```

### CODE REVIEW

**Review by sayboras — APPROVED** [review]
LGTM for file(s) owned by clustermesh team :heavy_check_mark:. 

I checked other files as well, but don't really have any comment.

File: pkg/maps/nat/per_cluster_nat.go [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+	perClusterNATMapMaxEntries   = cmtypes.ClusterIDMax + 1
+)
+
+// Global interface to interact with IPv4 and v6 NAT maps. We can choose the
+// implementation of this at startup time by choosing InitPerClusterNATMaps
+// or InitDummyPerClusterNATMaps for initialization.
+var PerClusterNATMaps PerClusterNATMapsIface
+
+// An interface to interact with the global map.
+type PerClusterNATMapsIface interface {
```
Comment (ldelossa) [inline pkg/maps/nat/per_cluster_nat.go]: I do not typically see Go interfaces called "Iface". Any better naming for this?

**Comment by ldelossa** [comment]
> However, the problem is we cannot extend the NAT map without breaking users' connections now. Thus, we created a new map-in-map containing the NAT maps for each cluster.

When it comes to utilizing this in the datapath, how do you see the NAT lookup process playing out? Will both NAT tables be checked, and what takes precedence.

**Review by ldelossa — APPROVED** [review]
Nothing concerning. A small nit about a name, which is ignorable. 

Did have a question on how this effects the normal NAT lookup flow in the datapath once utilized.

File: pkg/maps/nat/per_cluster_nat.go (hunk 1 above)
Comment (YutaroHayakawa) [reply pkg/maps/nat/per_cluster_nat.go]: Yep, Timo pointed out the same in this comment https://github.com/cilium/cilium/pull/22857#discussion_r1057240088. I'll rename it to `PerClusterNATMapper`

**Comment by YutaroHayakawa** [comment]
> Did have a question on how this effects the normal NAT lookup flow in the datapath once utilized.

Once it is utilized

1) It will start using SNAT for inter-cluster communication
2) The map instance will be selected depending on the destination cluster

By doing this, we can distinguish the flows like `10.0.0.1 => 10.1.0.1@1` and `10.0.0.1 => 10.1.0.1@2`.

**Comment by YutaroHayakawa** [comment]
Changes made
- Surround the per-cluster map definitions with `ENABLE_CLUSTER_AWARE_ADDRESSING` guard macro to avoid making it automatically.
- Renamed `PerClusterNATMapsIface` to `PerClusterNATMapper`

**Comment by YutaroHayakawa** [comment]
- ConformanceGKE: Timeout during resource cleanup. If retry doesn't help, we may need to extend timeout.
- ConformanceAKS: Failing with a new flake. Filed an issue https://github.com/cilium/cilium/issues/23070

**Comment by YutaroHayakawa** [comment]
ConformanceAKS failed with a known flake https://github.com/cilium/cilium/issues/20887. I believe this per-cluster NAT change doesn't make any difference with the cloud provider it runs. Let me make this ready-to-merge.

**Pull request merged**